AutoML with holdout dataset and improved descriptions

unmonoqueteclea opened the pull request:
In order to avoid any kind of overoptimistic evaluations in AutoML, we are
creating a holdout dataset from the beginning, different from the
validation dataset used internally in Feature Selection stage.

So this new holdout dataset is not used until the final evaluation
step.

OptiML does not receive a test-dataset any more, so it will perform
cross-validation. It makes no sense using a specific validation
dataset in the OptiML, because this dataset won't be provided by the
user, it will be sampled from the provided train-datasets.

I also improved the descriptions of some of the scripts as propsed by Guillem (/cc @guillembigml)

charleslparker commented:
@unmonoqueteclea - Thinking on this a bit more; I think it might still be a good idea to still hold out data for the OptiML that you do not use during feature selection.  This doesn't.have any impact on the final evaluation, but I think if you cross validate in the model on the same data you used for feature selection, you will end up with a bias in model selection towards models that can take advantage of that information leak (though it's not immediately clear to me how this might play out in the various types of models we offer).

This is closely related to the issue raised in the Hastie book.  As he says, all of the evaluations in the OptiML will be artificially higher if you cross validate in the OptiML.  That's easy to see.  What I'm wondering here is if the *relative quality* will also be different.  That is, will the evaluations in the OptiML be higher by different amounts in a way that will cause you to choose the wrong models for the fusion?

In my opinion, the odds of that happening are high enough where you should play it safe:  Split your training set 80/20, do feature selection on the 80% then run OpitML with the 80% as training data and 20% as validation.  Select the best models from the OptiML, fuse, and test on the provided test set.

unmonoqueteclea commented:
Thanks, @charleslparker, I added a new holdout dataset for OptiML as you suggested.
Ready to be reviewed, @mmerce.

mmerce commented on automl/automl-library/library.whizzml:
```diff
@@ -529,15 +536,13 @@
-                          (merge max-time {"test_dataset" validation})
-                          max-time)))
-      (create-optiml train params))))
+    (let (model-params (merge {"max_training_time" (if shallow 900 3600)
```
here do you want the user parameters to have precedence over these max_training_time and test_dataset?

unmonoqueteclea replied:
Yes

mmerce commented on automl/automl-script/metadata.json:
```diff
@@ -1,28 +1,28 @@
   "name": "AutoML",
-  "description":
-  "This script executes all the steps from the BigML's Automated Machine Learning pipeline",
+  "description":"These WhizzML script will let the user run a **fully automated\nMachine Learning pipeline** in BigML. It need the AutoML Library, more info [here](https://github.com/whizzml/examples/tree/master/automl).\n\n\nFrom a train, holdout and test datasets, the following tasks will\nbe automatically done:\n\n-  `Unsupervised Models Generation`: Creating the following\n  **unsupervised mo ...
```
It need**s** the AutoML...

mmerce commented on automl/automl-script/metadata.json:
```diff
@@ -1,28 +1,28 @@
       "type": "dataset-id",
       "default": "",
-      "description": "Validation dataset id (e.g. dataset/5d272205eba31d61920005cd)"
+      "description": "Hold-out dataset id, only to perform final evaluation"
```
Hold-out dataset ID. Used to evaluate the final Fusion model

mmerce commented on automl/automl-script/metadata.json:
```diff
@@ -1,28 +1,28 @@
       "type": "dataset-id",
       "default": "",
-      "description": "Test dataset id (e.g. dataset/5d272205eba31d61920005cd)"
+      "description": "Test dataset id, for scoring"
```
Test dataset ID. Inputs that will be scored using the final Fusion model

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -21,24 +21,28 @@ be automatically done:
 
 The **inputs** for the script are:
 
 * `train-dataset`: (dataset-id) Dataset id for the train dataset
```
dataset ID fro the train...

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -21,24 +21,28 @@ be automatically done:
+  provided, the script will expect an `automl-execution` and will use
+  its models as starting point. Useful when we only want to perform
+  scoring , using  a previous execution of AutoML.
+* `holdout-dataset`: (dataset-id) Dataset id for the holdout dataset
```
dataset ID

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -21,24 +21,28 @@ be automatically done:
+  scoring , using  a previous execution of AutoML.
+* `holdout-dataset`: (dataset-id) Dataset id for the holdout dataset
+  (e.g. dataset/5d272205ea31d61920005cd). It will be used **only at
+  the end of the workflow** after the model has been trained, to
```
... after the resulting fusion model has been created, to evaluate it.

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -21,24 +21,28 @@ be automatically done:
+  `train-dataset` will be set aside and used as `holdout-dataset`. The
+  parameter `holdout-rate` indicates the portion of the
+  `train-dataset` that is sampled in the `holdout-dataset`
 * `test-dataset`: (dataset-id) Dataset id for the test dataset
```
Dataset ID...

mmerce commented:
Let me see if I understand the present status:

- We allow the user to set 3 datasets: training, test and holdout
- We extend them with the features created from unsupervised learning models
- We split the training dataset to hold out a validation-dataset (used only in OptiML selection)
- We split the remaining training dataset to hold out a holdout if the user gives us a holdout rate (used only in the Fusion evaluation at the end)
- Now we have [training, validation, holdout, (and maybe) test].

Evaluations:
- In RFE we split again the training part to hold out a different validation dataset to do the feature selection.
- In OptiML we use the validation dataset held out in step 3 to do the type+parameters selection,
- We create a Fusion and an evaluation using the holdout-dataset
For this part, every time we run AutoML we get different splits, so non-deterministic behaviour, right? Which I assume is itended to be able to test consistent results over different executions.

Testing:
Iff there's a test-dataset:
- We create a batch-prediction for the test-dataset based on the previous Fusion

@unmonoqueteclea, please correct me if I'm wrong

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -21,24 +21,28 @@ be automatically done:
-  dataset instances. If a validation dataset is given, this script
-  will return an evaluation of the final model too.
+  dataset instances. The script will return an evaluation of the final
+  model too.
```
I would replace the last line of the previous paragraph and place a new item: 

- Fusion evaluation: Evaluating the Fusion using the data in the holdout dataset
- Fusion prediction: Scoring the input data provided in the test dataset (if set) using the final Fusion model.

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -95,14 +99,24 @@ overwritten by the corresponding input in `automl-execution` if this
-(train, validation or test) they are set as non-preferred fields.
+(train, holdout or test) they are set as non-preferred fields.
+
+**WARNING** In order to know your AutoML model performance, you should
```
The AutoML script creates partial models and evaluations both in the **RFE** and the **OptiML** selection processes. Those models and evaluations are using samples of the training data and independent validation datasets (extracted from the training data as totally complementary splits to avoid leakage between the Feature Selection and Model Selection steps). Therefore, only the final evaluation of the Fusion model on the holdout data can be used as an estimate for its performance.

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -111,23 +125,26 @@ The **outputs** for the script are:
-### From a train and a test dataset (and an optional validation dataset)
+### From a train and a test dataset (and an optional holdout dataset)
 In this case, the expected inputs for the script are the
 `train-dataset` and the `test-dataset`, with no `autml-execution`
```
automl-execution

mmerce commented on automl/automl-script/readme.md:
```diff
@@ -111,23 +125,26 @@ The **outputs** for the script are:
 input.  The objective field used in the script will be the one
 associated to the datasets used, so remember to choose them previously
-in both the training and validation datasets.
+in both the training and holdout datasets.
```
the training dataset (and holdout dataset when used).

mmerce commented on automl/automl-script/script.whizzml:
```diff
@@ -171,40 +170,39 @@
 
 
 ;; Obtains the most important fields (a list with their names)
 ;; from a train and an optional test dataset
```
remove: and an optional test dataset

mmerce commented on automl/automl-script/script.whizzml:
```diff
@@ -223,7 +221,7 @@
 ;; Creates the final fusion model with the top 5 models
```
docs says that you use 3 models in the Fusion

mmerce commented on fuzzy-norms/metadata.json:
```diff
@@ -1,7 +1,8 @@
 {
   "name": "Fuzzy Logic norms",
-  "description": "Computing fuzzy-logic T-norms and T-conorms as new dataset features",
+  "description": "In this package, you will find some utilities to apply fuzzy logic\nt-norms and t-conorms to datasets.\n\nYou can find all the information about the T-norms and T-conorms in\nthe following links:\n\n\n- [T-norms and T-conorms](https://en.wikipedia.org/wiki/T-norm)\n\n - [Construction of T-norms](https://en.wikipedia.org/wiki/Construction_of_t-norms)\n\nFields used as operands mu ...
```
Bellow -> Below

As you can see there are some norms that need a parameter, that...-> As you can see, there are some norms that need a parameter that...

It'ss dual -> It's dual

nilpotent minumum -> nilpotent minimum

unmonoqueteclea commented:
> Let me see if I understand the present status:
> 
>     * We allow the user to set 3 datasets: training, test and holdout
> 
>     * We extend them with the features created from unsupervised learning models
> 
>     * We split the training dataset to hold out a validation-dataset (used only in OptiML selection)
> 
>     * We split the remaining training dataset to hold out a holdout (used only in the Fusion evaluation at the end)
> 
>     * Now we have [training, validation, holdout, (and maybe) test].
> 
> 
> Evaluations:
> 
>     * In RFE we split again the training part to hold out a different validation dataset to do the feature selection.
> 
>     * In OptiML we use the validation dataset held out in step 3 to do the type+parameters selection,
> 
>     * We create a Fusion and an evaluation using the holdout-dataset
>       For this part, every time we run AutoML we get different splits, so non-deterministic behaviour, right? Which I assume is itended to be able to test consistent results over different executions.
> 
> 
> Testing:
> Iff there's a test-dataset:
> 
>     * We create a batch-prediction for the test-dataset based on the previous Fusion
> 
> 
> @unmonoqueteclea, please correct me if I'm wrong

Yes, all of this is mostly right. Only one detail. As you said, we need a holdout dataset to evaluate the final Fusion. This can be given by the user (it's one of the scripts input) but if not, the first thing we do it's obtaining it from the training dataset. The remaining data will be split again to obtain the validation dataset (used in OptiML). And yes, again in the Feature Selection step, the remaining train dataset is split. 

Thank you very much for your exhaustive review, I will check everything tomorrow.

unmonoqueteclea commented:
Ready. Thank you again, @mmerce
